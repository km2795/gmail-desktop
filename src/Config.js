"use strict";

require("dotenv").config();

const Config = {

  /* Root storage directory. */
  STORE_DIR: process.env.STORE_DIR,

  /* Logs storage directory. */
  LOG_DIR: process.env.LOG_DIR,

  /* File for storing UserList JSON. */
  USER_LIST_FILE: process.env.USER_LIST_FILE,

  /* Storage directory of the user. */
  USER_STORE_DIR: process.env.USER_STORE_DIR,

  /* User's email address. */
  USER_ID: "",

  /* Current user. */
  User: "",

  /* Client's secret credentials. */
  CLIENT_SECRET: process.env.CLIENT_SECRET,

  /* Label's file location. */
  LABELS_STORE: "",

  /* History ID file location. */
  HISTORY_ID_STORE: "",

  /* Message fetching interval time. */
  FETCH_INTERVAL: process.env.FETCH_INTERVAL,

  /*
   * To avoid multiple instances of fetchMessages() to
   * run concurrently.
   * true - Another instance is running.
   * false - No other instance is running.
   */
  FetchingMessages: false,

  /* Current User's Mailbox labels. */
  Labels: {},

  /* Logger object. */
  Logger: null,

  /* Color for individual Mailbox label's badge. */
  LABEL_COLORS: {
    "inbox": "red",
    "trash": "black",
    "important": "cyan",
    "chat": "lime",
    "sent": "black",
    "draft": "maroon",
    "spam": "yellow",
    "starred": "lime",
    "unread": "teal",
    "label_1": "salmon",
    "label_5": "indianred",
    "label_6": "lightcoral",
    "label_2": "beige",
    "category_updates": "orange",
    "category_social": "blue",
    "category_forums": "purple",
    "category_promotions": "green",
    "category_personal": "light-green"
  }
};

export default Config;
