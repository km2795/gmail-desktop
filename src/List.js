function List (buffer, delimiter) {
  // eslint-disable-next-line no-empty-pattern
  const self = {} = this;

  /*
   * If no parameter is passed.
   * Default constructor kind-of.
   */
  if (buffer == undefined) {
    return;
  }

  // Temporary.
  let list = [];

  // Check the type of object.
  switch (Object.prototype.toString.call(buffer)) {
    // JSON object.
    case "[object Object]":
      this.load(buffer);
      return;

    // Buffer.
    case "[object Uint8Array]":
      list = buffer.toString().split(delimiter);
      break;

    // Array.
    case "[object Array]":
      list = buffer;
      break;

    // String.
    case "[object String]":
      list = buffer.split(delimiter);
      break;
  }

  // Create JSON from array with numeric indices.
  list.map((item, i) => {
    self[i++] = item;
  });
}

List.prototype.load = function (obj) {
  const self = this;
  const keys = Object.keys(obj);
  keys.map((item) => {
    self.add(item, obj[item]);
  });
}

/**
 * Add key/value to the List.
 *
 * @param {string} key - Key of the entry.
 * @param {any} value - Value of the entry.
 *
 * @returns {}
 */
List.prototype.add = function (key, value) {
  this[key] = value;
};

/**
 * Remove entry from the List.
 *
 * @param {string} key - Key of the entry to be removed.
 *
 * @returns {}
 */
List.prototype.remove = function (key) {
  delete this[key];
};

/**
 * Get the object value from the List with the help of the ID.
 *
 * @param {string} key - Key of the entry whose value is to be retrieved.
 *
 * @returns {any} - Value.
 */
List.prototype.get = function (key) {
  return this[key];
};

/**
 * Return the number of key/value pairs in the List.
 *
 * @param {}
 *
 * @returns {number} - Size of the List.
 */
List.prototype.size = function () {
  return this.getKeys().length;
};

/**
 * Return a list containing all the keys of the List.
 *
 * @param {}
 *
 * @returns {[]} - List containing all the keys.
 */
List.prototype.getKeys = function () {
  return Object.keys(this);
};

/**
 * Returns a list containing all the values of the List.
 *
 * @param {}
 *
 * @returns {[]} - List containing all the values.
 */
List.prototype.getValues = function () {
  return this.getKeys().map((i) => this[i]);
};

export default List;
