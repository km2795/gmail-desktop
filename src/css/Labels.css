.message-labels {
  margin: 0px;
  text-align: right;
}
