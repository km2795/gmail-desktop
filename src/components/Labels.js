import React from "react";
import Label from "./Label";
import "../css/Labels.css";

const Labels = ({ threadLabels, LABEL_COLORS, LABELS }) => (
  <p className="message-labels">{
    threadLabels.map((label, i) => (
      (label !== "UNREAD")
        ? (<Label key={i++}
          color={
            LABEL_COLORS[label.toLowerCase()]
              ? LABEL_COLORS[label.toLowerCase()]
              : "white"
          }
          label={LABELS.get(label).name}
          userLabel={LABELS.get(label).type === "user"}
        />)
        : null
    ))
  }</p>
);

export default Labels;
