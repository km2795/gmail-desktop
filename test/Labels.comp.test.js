import React from "react";
import renderer from "react-test-renderer";
import Config from "../src/Config";
import List from "../src/List";
import Labels from "../src/components/Labels";

const TEST_LABELS = new List({
  "CATEGORY_PERSONAL": {
    "id": "CATEGORY_PERSONAL",
    "name": "CATEGORY_PERSONAL",
    "type": "system"
  },
  "CATEGORY_SOCIAL": {
    "id": "CATEGORY_SOCIAL",
    "name": "CATEGORY_SOCIAL",
    "type": "system"
  },
  "Label_1": {
    "id": "Label_1",
    "name": "[Imap]/Trash",
    "type": "user"
  },
  "CATEGORY_FORUMS": {
    "id": "CATEGORY_FORUMS",
    "name": "CATEGORY_FORUMS",
    "type": "system"
  },
  "IMPORTANT": {
    "id": "IMPORTANT",
    "name": "IMPORTANT",
    "messageListVisibility": "hide",
    "labelListVisibility": "labelShow",
    "type": "system"
  },
  "CATEGORY_UPDATES": {
    "id": "CATEGORY_UPDATES",
    "name": "CATEGORY_UPDATES",
    "type": "system"
  },
  "CHAT": {
    "id": "CHAT",
    "name": "CHAT",
    "messageListVisibility": "hide",
    "labelListVisibility": "labelHide",
    "type": "system"
  },
  "SENT": {
    "id": "SENT",
    "name": "SENT",
    "messageListVisibility": "hide",
    "labelListVisibility": "labelShow",
    "type": "system"
  },
  "INBOX": {
    "id": "INBOX",
    "name": "INBOX",
    "messageListVisibility": "hide",
    "labelListVisibility": "labelShow",
    "type": "system"
  },
  "TRASH": {
    "id": "TRASH",
    "name": "TRASH",
    "messageListVisibility": "hide",
    "labelListVisibility": "labelHide",
    "type": "system"
  },
  "CATEGORY_PROMOTIONS": {
    "id": "CATEGORY_PROMOTIONS",
    "name": "CATEGORY_PROMOTIONS",
    "type": "system"
  },
  "DRAFT": {
    "id": "DRAFT",
    "name": "DRAFT",
    "messageListVisibility": "hide",
    "labelListVisibility": "labelShow",
    "type": "system"
  },
  "SPAM": {
    "id": "SPAM",
    "name": "SPAM",
    "messageListVisibility": "hide",
    "labelListVisibility": "labelHide",
    "type": "system"
  },
  "STARRED": {
    "id": "STARRED",
    "name": "STARRED",
    "messageListVisibility": "hide",
    "labelListVisibility": "labelShow",
    "type": "system"
  },
  "UNREAD": {
    "id": "UNREAD",
    "name": "UNREAD",
    "type": "system"
  },
  "Label_2222222222222222222": {
    "id": "Label_2222222222222222222",
    "name": "Archive_Payments",
    "type": "user"
  }
});

test("Should render labels of a thread", () => {
  const tree = renderer
    .create(<Labels 
      threadLabels={["CATEGORY_UPDATES", "INBOX", "CHAT", "Label_2222222222222222222"]}
      LABEL_COLORS={Config.LABEL_COLORS}
      LABELS={TEST_LABELS} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
