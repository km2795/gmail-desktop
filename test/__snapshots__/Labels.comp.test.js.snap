// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render labels of a thread 1`] = `
<p
  className="message-labels"
>
  <span
    className="
    badge badge-orange 
    message-label-badge 
    "
  >
    UPDATES
  </span>
  <span
    className="
    badge badge-red 
    message-label-badge 
    "
  >
    INBOX
  </span>
  <span
    className="
    badge badge-lime 
    message-label-badge 
    "
  >
    CHAT
  </span>
  <span
    className="
    badge badge-white 
    message-label-badge 
    user-label"
  >
    Archive_Payments
  </span>
</p>
`;
